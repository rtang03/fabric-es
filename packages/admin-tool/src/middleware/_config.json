{
  "trade-network": {
    "orderer": {
      "mspid": "OrdererMSP",
      "url": "grpcs://localhost:7050",
      "hostname": "orderer.example.com",
      "tls_cacerts": "../../../../../network/hosts/org0/orderer/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem"
    },
    "org1": {
      "name": "org1",
      "mspid": "Org1MSP",
      "ca": {
        "url": "https://localhost:5054",
        "name": "rca-org1"
      },
      "peer0": {
        "requests": "grpcs://localhost:7051",
        "events": "grpcs://localhost:7053",
        "hostname": "peer0.org1.example.com",
        "tls_cacerts": "../../../../network/hosts/org1/peer0/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem"
      },
      "peer1": {
        "requests": "grpcs://localhost:8051",
        "events": "grpcs://localhost:8053",
        "hostname": "peer1.org1.example.com",
        "tls_cacerts": "../../../../network/hosts/org1/peer1/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem"
      }
    },
    "org2": {
      "name": "org2",
      "mspid": "Org2MSP",
      "ca": {
        "url": "https://localhost:5055",
        "name": "rca-org2"
      },
      "peer0": {
        "requests": "grpcs://localhost:9051",
        "events": "grpcs://localhost:9053",
        "hostname": "peer0.org2.example.com",
        "tls_cacerts": "../../../../network/hosts/org2/peer0/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem"
      },
      "peer1": {
        "requests": "grpcs://localhost:10051",
        "events": "grpcs://localhost:10053",
        "hostname": "peer1.org2.example.com",
        "tls_cacerts": "../../../../network/hosts/org2/peer1/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem"
      }
    }
  }
}
