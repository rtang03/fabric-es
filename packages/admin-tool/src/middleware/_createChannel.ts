// import * as Client from 'fabric-client';
// import {
//   newCryptoKeyStore,
//   newCryptoSuite,
//   newDefaultKeyValueStore
// } from 'fabric-client';
// import { readFileSync } from 'fs';
// import { join } from 'path';
// import '../env';
// import { Context } from './types';
// import { clientUtils } from './utils';
//
// export const createChannel: (
//   channelName: string,
//   org: string,
//   context?: Context
// ) => Promise<{ status: any }> = async (
//   channelName,
//   org,
//   context = {
//     kvStorePath: process.env.PATH_TO_CERTS,
//     networkId: process.env.NETWORK_ID || 'trade-network',
//     pathToChannelTx: process.env.PATH_TO_CHANNEL_CONFIG,
//     pathToNetwork: process.env.PATH_TO_NETWORK,
//     pathToNetworkConfig: process.env.PATH_TO_NETWORK_CONFIG
//   }
// ) => {
//   const { kvStorePath, pathToChannelTx, pathToNetworkConfig } = context;
//   const path = kvStorePath + `_${org}`;
//   const { client, config, signatures } = await newDefaultKeyValueStore({ path })
//     .then(store => {
//       Client.addConfigFile(join(__dirname, pathToNetworkConfig));
//       Client.setConfigSetting(
//         'key-value-store',
//         'fabric-client/lib/impl/FileKeyValueStore.js'
//       );
//       const cryptoSuite = newCryptoSuite();
//       cryptoSuite.setCryptoKeyStore(newCryptoKeyStore({ path }));
//       const client = new Client();
//       client.setStateStore(store);
//       client.setCryptoSuite(cryptoSuite);
//       console.log('Successfully extracted the config from the configtx');
//       return client;
//     })
//     .then(async client => {
//       const signatures = [];
//       const envelope = readFileSync(
//         join(__dirname, pathToChannelTx, `${channelName}.tx`)
//       );
//       const config = client.extractChannelConfig(envelope);
//       for (const org of clientUtils(client, context).peerOrgs)
//         signatures.push(
//           // Enrol admin AND Sign Channel Config
//           await clientUtils(client, context)
//             .enrolAdmin(org)
//             .then(() => client.signChannelConfig(config))
//         );
//       return { signatures, client, config };
//     });
//
//   const channel = client.newChannel(channelName);
//   const orderer = clientUtils(client, context).getOrderer();
//   channel.addOrderer(orderer);
//
//   return await channel.getGenesisBlock().then(
//     () => {
//       console.log(`Got genesis block. Channel: ${channelName} already exists`);
//       return { status: 'SUCCESS' };
//     },
//     async () =>
//       await client
//         .createChannel({
//           config,
//           signatures,
//           name: channelName,
//           orderer,
//           txId: client.newTransactionID()
//         })
//         .then(result => {
//           console.log(`Channel ${channelName} does not exist yet`);
//           if (result.status === 'SUCCESS') {
//             // console.log(result);
//             return result;
//           } else throw new Error('Failed to create the channel.');
//         })
//   );
// };
