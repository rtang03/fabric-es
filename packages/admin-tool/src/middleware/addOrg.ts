[ [ { version: 0,
  timestamp: null,
  response: [Object],
  payload: <Buffer >,
  endorsement: null,
  peer: [Object] },
  { version: 0,
    timestamp: null,
    response: [Object],
    payload: <Buffer >,
    endorsement: null,
    peer: [Object] } ],
  [ { version: 0,
    timestamp: null,
    response: [Object],
    payload: <Buffer >,
    endorsement: null,
    peer: [Object] },
    { version: 0,
      timestamp: null,
      response: [Object],
      payload: <Buffer >,
      endorsement: null,
      peer: [Object] } ] ]
