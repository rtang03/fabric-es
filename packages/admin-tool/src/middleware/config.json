{
  "trade-network": {
    "orderer": {
      "mspid": "TradeOrdererMSP",
      "url": "grpcs://localhost:7050",
      "hostname": "orderer.example.com",
      "tls_cacerts": "../../../../network/hosts/org0/orderer/tls-msp/tlscacerts/tls-0-0-0-0-5052.pem"
    },
    "org1": {
      "name": "org1",
      "mspid": "Org1MSP",
      "ca": {
        "url": "https://localhost:5054",
        "name": "rca-org1"
      },
      "peer0": {
        "requests": "grpcs://localhost:7051",
        "events": "grpcs://localhost:7053",
        "server-hostname": "peer0.org1.example.com",
        "tls_cacerts": "../network/crypto-config/peerOrganizations/exporterorg.trade.com/peers/peer0.exporterorg.trade.com/msp/tlscacerts/tlsca.exporterorg.trade.com-cert.pem"
      },
      "peer1": {
        "requests": "grpcs://localhost:8051",
        "events": "grpcs://localhost:8053",
        "server-hostname": "peer1.org1.example.com",
        "tls_cacerts": "../network/crypto-config/peerOrganizations/exporterorg.trade.com/peers/peer0.exporterorg.trade.com/msp/tlscacerts/tlsca.exporterorg.trade.com-cert.pem"
      }
    },
    "org2": {
      "name": "org2",
      "mspid": "Org2MSP",
      "ca": {
        "url": "https://localhost:5055",
        "name": "rca-org2"
      },
      "peer0": {
        "requests": "grpcs://localhost:9051",
        "events": "grpcs://localhost:9053",
        "server-hostname": "peer0.org2.example.com",
        "tls_cacerts": "../network/crypto-config/peerOrganizations/importerorg.trade.com/peers/peer0.importerorg.trade.com/msp/tlscacerts/tlsca.importerorg.trade.com-cert.pem"
      },
      "peer1": {
        "requests": "grpcs://localhost:10051",
        "events": "grpcs://localhost:10053",
        "server-hostname": "peer1.org2.example.com",
        "tls_cacerts": "../network/crypto-config/peerOrganizations/importerorg.trade.com/peers/peer0.importerorg.trade.com/msp/tlscacerts/tlsca.importerorg.trade.com-cert.pem"
      }
    }
  }
}
