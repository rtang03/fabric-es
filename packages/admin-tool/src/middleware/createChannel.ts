import * as Client from 'fabric-client';
import { readFileSync } from 'fs';
import { join } from 'path';
import '../env';
import { Context } from './types';
import { connectionProfile, enrollAdmin, getClientForOrg } from './utils';

export const createChannel: (
  channelName: string,
  context?: Context
) => Promise<any> = async (
  channelName,
  context = {
    connProfileNetwork: process.env.PATH_TO_CONNECTION_PROFILE,
    pathToChannelTx: process.env.PATH_TO_CHANNEL_CONFIG,
    pathToNetwork: process.env.PATH_TO_NETWORK
  }
) => {
  const { pathToChannelTx, connProfileNetwork } = context;
  const client: Client = await getClientForOrg(connProfileNetwork);

  // create orderer
  const { getOrderer, getOrganizations } = await connectionProfile(context);
  const { url, tlsCACertsPem, hostname } = getOrderer();
  const { getOrgs } = getOrganizations();
  const orderer = client.newOrderer(url, {
    pem: tlsCACertsPem,
    'ssl-target-name-override': hostname
  });
  const channel = client.newChannel(channelName);
  channel.addOrderer(orderer);

  // enrol all org's admins, and sign channel configuration
  const config = client.extractChannelConfig(
    readFileSync(join(__dirname, pathToChannelTx, `${channelName}.tx`))
  );
  const signatures = [];
  for (const { orgName } of getOrgs()) {
    const admin = await getClientForOrg(connProfileNetwork);
    signatures.push(
      await enrollAdmin(admin, orgName, context).then(() =>
        admin.signChannelConfig(config)
      )
    );
  }
  return channel.getGenesisBlock().then(
    () => {
      console.log(`Got genesis block. Channel: ${channelName} already exists`);
      return { status: 'SUCCESS' };
    },
    async () =>
      await client
        .createChannel({
          config,
          signatures,
          name: channelName,
          orderer: hostname,
          txId: client.newTransactionID(true)
        })
        .then(result => {
          console.log(`Channel ${channelName} does not exist yet`);
          if (result.status === 'SUCCESS') {
            return result;
          } else throw new Error('Failed to create the channel.');
        })
  );
};
