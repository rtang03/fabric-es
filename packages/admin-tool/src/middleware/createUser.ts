import { FileSystemWallet, Gateway, X509WalletMixin } from 'fabric-network';
import { Context } from './types';
import { getClientForOrg } from './utils';

export const createUser: (
  enrollmentID: string,
  enrollmentSecret: string,
  context?: Context
) => Promise<any> = async (
  enrollmentID,
  enrollmentSecret,
  { connectionProfile, wallet } = {
    connectionProfile: process.env.PATH_TO_CONNECTION_PROFILE,
    wallet: new FileSystemWallet('./wallet')
  }
) => {
  const adminExist: boolean = await wallet.exists('admin');
  if (!adminExist) {
    throw new Error('No admin in the wallet; enroll admin before retrying');
  }

  return (await wallet.exists(enrollmentID))
    ? {
        status: 'SUCCESS',
        message: `Identity for ${enrollmentID} already exists in the wallet`
      }
    : getClientForOrg(connectionProfile).then(async admin => {
        const gateway = new Gateway();
        await gateway.connect(admin, {
          wallet,
          identity: 'admin',
          discovery: { enabled: true, asLocalhost: true }
        });
        const ca = gateway.getClient().getCertificateAuthority();
        await ca.register(
          {
            enrollmentID,
            enrollmentSecret,
            affiliation: '',
            maxEnrollments: -1,
            role: 'client'
          },
          gateway.getCurrentIdentity()
        );
        await ca
          .enroll({
            enrollmentID,
            enrollmentSecret
          })
          .then(
            async ({ key, certificate }) =>
              await wallet.import(
                enrollmentID,
                X509WalletMixin.createIdentity(
                  admin.getMspid(),
                  certificate,
                  key.toBytes()
                )
              )
          );
        return {
          status: 'SUCCESS',
          message: `Successfully register & enroll ${enrollmentID}; and import into the wallet`
        };
      });
};
