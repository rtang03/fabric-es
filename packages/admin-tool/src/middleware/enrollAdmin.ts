import { FileSystemWallet, X509WalletMixin } from 'fabric-network';
import '../env';
import { Context } from './types';
import { getCAServices, getClientForOrg } from './utils';

export const enrollAdmin: (
  enrollmentID: string,
  enrollmentSecret: string,
  url: string,
  orgName: string,
  context?: Context
) => Promise<any> = async (
  enrollmentID,
  enrollmentSecret,
  url,
  orgName,
  context = {
    connProfileNetwork: process.env.PATH_TO_CONNECTION_PROFILE,
    fabricNetwork: process.env.PATH_TO_NETWORK,
    wallet: new FileSystemWallet(process.env.WALLET)
  }
) =>
  getClientForOrg(context.connProfileNetwork).then(async admin => {
    const ca = await getCAServices(admin, url, orgName, context);
    return context.wallet.exists('admin').then(async exist =>
      exist
        ? {
            status: 'SUCCESS',
            message: 'Identity for "admin" already exists in the wallet'
          }
        : await ca
            .enroll({
              enrollmentID,
              enrollmentSecret
            })
            .then(
              async ({ key, certificate }) =>
                await context.wallet.import(
                  'admin',
                  X509WalletMixin.createIdentity(
                    admin.getMspid(),
                    certificate,
                    key.toBytes()
                  )
                )
            )
            .then(() => ({
              status: 'SUCCESS',
              message: 'Successfully enroll admin and import into the wallet'
            }))
    );
  });
