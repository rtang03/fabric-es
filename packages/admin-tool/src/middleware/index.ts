export * from './createChannel';
export * from './joinChannel';
export * from './installChaincode';
export * from './instantiateChaincode';
export * from './getInfo';
export * from './registerUser';
export * from './enrollAdmin';
export * from './enrollCAAdmin';
export * from './identity';
