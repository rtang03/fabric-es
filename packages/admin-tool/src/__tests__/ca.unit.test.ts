import { registerUser } from '../middleware';

const enrollmentID = `user${Math.floor(Math.random() * 100)}@test.com`;
const secret = 'testpw';
const caUrl = 'https://0.0.0.0:5054';
const orgName = 'Org1';

describe('Fabric CA unit test', () => {
  it('should register', async () =>
    await registerUser(enrollmentID, secret, caUrl, orgName).then(result =>
      expect(result).toBe(secret)
    ));
});
