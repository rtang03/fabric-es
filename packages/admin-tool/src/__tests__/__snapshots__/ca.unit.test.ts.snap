// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fabric CA unit test should enrol Org1 Admin 1`] = `
Object {
  "message": "Successfully enroll admin and import into the wallet",
  "status": "SUCCESS",
}
`;

exports[`Fabric CA unit test should enrol Org1 Admin 2`] = `
Object {
  "message": "Identity for \\"admin\\" already exists in the wallet",
  "status": "SUCCESS",
}
`;
